src/node_jsvmapi.cc

```diff
@@ -465,22 +465,109 @@ namespace v8impl {
+  RETURN_STATUS_IF_FAIL((arg), napi_invalid_arg)
+
+#define CHECK_MAYBE_EMPTY(maybe, status)                                \
+  RETURN_STATUS_IF_FAIL(!((maybe).IsEmpty()), (status))
+
+#define CHECK_MAYBE_NOTHING(maybe, status)                              \
+  RETURN_STATUS_IF_FAIL(!((maybe).IsNothing()), (status))
+
+#define NAPI_PREAMBLE(e)                                                \
+  do {                                                                  \
+    RETURN_STATUS_IF_FAIL(v8impl::TryCatch::lastException().IsEmpty(),  \
+      napi_pending_exception);                                          \
+    napi_clear_last_error_code();                                       \
+    v8impl::TryCatch tryCatch(v8impl::V8IsolateFromJsEnv((e)));         \
+  } while(0)
+
+#define CHECK_TO_OBJECT(result, o)                                      \
+  do {                                                                  \
+    auto obj_maybe =                                                    \
+      v8impl::V8LocalValueFromJsValue((o))->ToObject(context);          \
+    CHECK_MAYBE_EMPTY(obj_maybe, napi_object_expected);                 \
+    result = obj_maybe.ToLocalChecked();                                \
+  } while(0)
+
+#define CHECK_NEW_FROM_UTF8_LEN(result, str, len)                       \
+  do {                                                                  \
+    auto name_maybe = v8::String::NewFromUtf8(isolate, (str),           \
+      v8::NewStringType::kInternalized, len);                           \
+    CHECK_MAYBE_EMPTY(name_maybe, napi_generic_failure);                \
+    result = name_maybe.ToLocalChecked();                               \
+  } while(0)
+
+#define CHECK_NEW_FROM_UTF8(result, str)                                \
+    CHECK_NEW_FROM_UTF8_LEN((result), (str), -1)
+
+   // Static last error returned from napi_get_last_error_info
+napi_extended_error_info static_last_error;
+
+// Warning: Keep in-sync with napi_status enum
+const char* error_messages[] = {
+  NULL,
+  "Invalid pointer passed as argument",
+  "An object was expected",
+  "A string was expected",
+  "A function was expected",
+  "Unknown failure",
+  "An exception is pending"
+};
+
+void napi_clear_last_error_code() {
+  // TODO: Should this also clear engine_specific fields?
+  static_last_error.error_code = napi_ok;
+}
+
+const napi_extended_error_info* napi_get_last_error_info() {
+  static_assert(sizeof(error_messages) / sizeof(*error_messages) == napi_status_last);
+  assert(static_last_error.error_code < napi_status_last);
 
-napi_env napi_get_current_env() {
-  return v8impl::JsEnvFromV8Isolate(v8::Isolate::GetCurrent());
+  // Wait until someone requests the last error information to fetch the error message string
+  static_last_error.error_message = error_messages[static_last_error.error_code];
+
+  return &static_last_error;
+}
+
+void napi_clear_extended_error_info() {
```

Comment:
I think it would be simpler to combine `napi_clear_extended_error_info()` and `napi_clear_last_error_code()` in a single `napi_clear_last_error()` that clears everything. I can't think of a scenario where it would be necessary to clear only one part or the other.